### ISSUE TITLE
WSL 2: tmux zombies not reaped without shell parent

### ISSUE THREAD

**Issue by cerebrate** [microsoft/WSL#4138]
* Your Windows build number:  18917

* What you're doing and what's happening:

Running tmux under WSL 2. Initially, this was noticed using the

`exec dbus-run-session -- /usr/bin/tmux -l`

command in .bash_login I use to start tmux by default, but a simple

`wsl tmux` from the command prompt or `exec tmux` from bash works just as well.

* What's wrong / what should be happening instead:

When tmux is run in any of the above manners, i.e., the parent process of tmux is init, tmux's zombie children aren't reaped; ps -ef will show a

`avatar   14125 13594  0 13:53 pts/0    00:00:00 [tmux: client] <defunct>`

line and possibly a number of other defunct processes, including sh processes, depending on the tmux configuration. These zombie processes should be (and were, under WSL 1) reaped.

Note that these zombie processes do **not** remain when tmux is run from bash, or where a shell exists as a parent process, parent-of-parent process, etc., only where tmux is a direct child of init. While this provides a workaround, of course, it should not be necessary.

**Comment by cerebrate** [comment]
I'm mentioning this here because it's (I think) likely related, though it occurs even when tmux does have a shell n-parent. I use the [tmux-yank plugin](https://github.com/tmux-plugins/tmux-yank) for easy copying from WSL, which was working fine in WSL 1.

In 2, however, I end up with a couple of frozen non-zombie processes every time I copy some text:

```
avatar   20000  0.0  0.0   4280   744 ?        S    14:11   0:00 sh -c clip.exe
avatar   20015  0.0  0.0    768     4 ?        S    14:11   0:00 /tools/init /mnt/c/WINDOWS/system32/clip.exe
avatar   20768  0.0  0.0   4280   708 ?        S    14:17   0:00 sh -c clip.exe
avatar   20783  0.0  0.0    768     4 ?        S    14:17   0:00 /tools/init /mnt/c/WINDOWS/system32/clip.exe
```

Manually piping to the same command, i.e., `echo test | sh -c clip.exe` works, however.

**Comment by benhillis** [comment]
Could you share "ps -auxf" so I can see where the zombie children have been parented?  It's possible something in the init chain is not reaping zombies as it should.

**Comment by cerebrate** [comment]
Can do.

```
$ ps -auxf
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         1  0.0  0.0   1680  1256 ?        Sl   08:55   0:00 /init
root         9  0.0  0.0    864    84 ?        S    08:55   0:00 /init
root        15  0.0  0.0  56908  6644 ?        Ss   08:55   0:00  \_ /lib/systemd/systemd --system
root        27  0.0  0.0  46096  6980 ?        Ss   08:55   0:00  |   \_ /lib/systemd/systemd-journald
root        39  0.0  0.0  43972  3880 ?        Ss   08:55   0:00  |   \_ /lib/systemd/systemd-udevd
root        64  0.0  0.0 250116  2904 ?        Ssl  08:55   0:00  |   \_ /usr/sbin/rsyslogd -n
message+    66  0.0  0.0  45132  3964 ?        Ss   08:55   0:00  |   \_ /usr/bin/dbus-daemon --system --address=systemd: --nofork --nopidfile --systemd-activation
daemon     163  0.0  0.0  27968  2128 ?        Ss   08:56   0:00  |   \_ /usr/sbin/atd -f
root       164  0.0  0.0  27996  2408 ?        Ss   08:56   0:00  |   \_ /usr/sbin/cron -f
root       166  0.0  0.2 588288 45356 ?        Ssl  08:56   0:07  |   \_ /usr/bin/containerd
root       168  0.1  0.4 694208 68632 ?        Ssl  08:56   0:23  |   \_ /usr/bin/dockerd -H fd:// --containerd=/run/containerd/containerd.sock
root       852  0.0  0.1 341216 15344 ?        Ssl  09:01   0:00  |   \_ /usr/lib/snapd/snapd
root      1096  0.0  0.0 299804  8752 ?        Ssl  09:12   0:00  |   \_ /usr/lib/policykit-1/polkitd --no-debug
Debian-+  3127  0.0  0.0  56156  3496 ?        Ss   09:28   0:00  \_ /usr/sbin/exim4 -bd -q30m
root       238  0.0  0.0    864    84 ?        S    08:56   0:02 /init
avatar     287  0.0  0.0  11092  2692 ?        Ss   08:56   0:00  \_ ssh-agent
root     24595  0.0  0.0   1680   872 ?        Ss   14:36   0:00 /init
root     24597  0.0  0.0   1680   872 ?        S    14:36   0:00  \_ /init
avatar   24600  0.0  0.0  21684  3168 pts/0    Ss+  14:36   0:00      \_ tmux
avatar   24601  0.0  0.0      0     0 pts/0    Z+   14:36   0:00      |   \_ [tmux: client] <defunct>
avatar   24602  1.1  0.0  30472  3800 ?        Ss   14:36   0:00      \_ tmux
avatar   24603  0.0  0.0      0     0 ?        Z    14:36   0:00          \_ [sh] <defunct>
avatar   24604  2.4  0.0  43916  6420 pts/1    Ss   14:36   0:00          \_ -zsh
avatar   25344  0.0  0.0  23680  7188 pts/1    S+   14:36   0:00          |   \_ /usr/bin/python /usr/bin/grc ps -auxf
avatar   25345  0.0  0.0  38312  3356 pts/1    R+   14:36   0:00          |       \_ ps -auxf
avatar   25346  0.0  0.0  20328  4196 pts/1    R+   14:36   0:00          |       \_ /usr/bin/python /usr/bin/grcat conf.ps
avatar   25139  0.0  0.0      0     0 ?        Z    14:36   0:00          \_ [sh] <defunct>
avatar   25140  0.0  0.0      0     0 ?        Z    14:36   0:00          \_ [sh] <defunct>
```

**Comment by benhillis** [comment]
@cerebrate - Thanks!  I'm utilizing the PR_SET_CHILD_SUBREAPER to track process lifetimes for daemons etc, I probably have a subtle issue there.  Digging into this now.

**Comment by therealkenc** [comment]
> I probably have a subtle issue there.

This is screaming never really explored or conclusively resolved #3914. The spinning and curiously accumulating tmux processes "I noticed sometimes my computer lags" part, not the by-design part. Maybe.

**Comment by markuszeller** [comment]
I am having the same issue. A ZSH process is left over and consumes about 20-25% CPU usage per process. This is on two Windows 10 computers, Version 10.0.18362 Build 18362.
